apisix/server_info.lua

```diff
@@ -0,0 +1,168 @@
+local internal_status = ngx.shared.internal_status
+
+local _M = {}
+
+if not internal_status then
+    error("lua_shared_dict \"internal_status\" not configured")
+end
+
+
+local function is_privileged()
+    local process_type = require("ngx.process").type()
+    return process_type == "privileged agent" or process_type == "single"
+end
+
+-- server information will be saved into shared memory only if the key
+-- "server_info" not exist if excl is true.
+local function save(data, excl)
+    local handler = excl and internal_status.add or internal_status.set
+
+    local ok, err = handler(internal_status, "server_info", data)
+    if not ok then
+        if excl and err == "exists" then
+            return true
+        end
+
+        return nil, err
+    end
+
+    return true
+end
+
+
+local function encode_and_save(server_info, excl)
+    local data, err = core.json.encode(server_info)
+    if not data then
+        return nil, err
+    end
+
+    return save(data, excl)
+end
+
+
+local function report()
+    local server_info, err = _M.get()
+    if not server_info then
+        core.log.error("failed to get server_info: ", err)
+        return nil, err
+    end
+
+    if server_info.etcd_version == "unknown" then
+        local res, err = core.etcd.server_version()
+        if not res then
+            core.log.error("failed to fetch etcd version: ", err)
+            return nil, err
+
+        elseif type(res.body) ~= "table" then
+            core.log.error("failed to fetch etcd version: bad version info")
+            return nil, "bad etcd version info"
+        else
+            server_info.etcd_version = res.body.etcdcluster
+        end
+    end
+
+    server_info.last_report_time = ngx_time()
+
+    local data, err = core.json.encode(server_info)
+    if not data then
+        core.log.error("failed to encode server_info: ", err)
+        return nil, err
+    end
+
+    local key = "/data_plane/server_info/" .. server_info.id
+    local ok, err = core.etcd.set(key, data, 180)
+    if not ok then
+        core.log.error("failed to report server info to etcd: ", err)
+        return nil, err
+    end
+
+    local ok, err = save(data, false)
+    if not ok then
+        core.log.error("failed to encode and save server info: ", err)
+        return nil, err
+    end
+end
+
+
+local function uninitialized_server_info()
+    return {
+        etcd_version     = "unknown",
+        hostname         = core.utils.gethostname(),
+        id               = core.id.get(),
+        version          = core.version.VERSION,
+        up_time          = ngx_time() - boot_time,
+        last_report_time = -1,
+    }
+end
+
+
+function _M.init_worker()
+    if not is_privileged() then
+        return
+    end
+
+    local ok, err = encode_and_save(uninitialized_server_info(), true)
+    if not ok then
+        core.log.error("failed to encode and save server info: ", err)
+    end
+
+    local opts = {
+        check_interval = 5, -- in seconds
```

Comment:
`5` seconds is short, I think 10 mins should be good

Reply:
and we should allow the user to specify this field

Reply:
I still think 10 mins is too long, also for now, all the basic server info are static (except the `up_time` and `last_report_time`), a relative short period is more suitable so we can easily judge whether a node is healthy or unstable   when we are viewing the server info on Dashboard.

What about 1 min or 2 mins?